### PR TITLE
Add LTX 2.0 Video Pipelines

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR adds pipelines for the LTX 2.0 video generation model ([code](https://github.com/Lightricks/LTX-2), [weights](https://huggingface.co/Lightricks/LTX-2)). LTX 2.0 is an audio-video foundation model that generates videos with synced audio; it supports generation tasks such as text-to-video (T2V), text-image-to-video (TI2V), and more.

An example usage script for I2V is as follows:

```python
import torch
from diffusers.pipelines.ltx2 import LTX2ImageToVideoPipeline
from diffusers.pipelines.ltx2.export_utils import encode_video
from diffusers.utils import load_image

pipe = LTX2ImageToVideoPipeline.from_pretrained("Lightricks/LTX-2", torch_dtype=torch.bfloat16)
pipe.enable_model_cpu_offload()

image = load_image(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/astronaut.jpg"
)
prompt = "An astronaut hatches from a fragile egg on the surface of the Moon, the shell cracking and peeling apart in gentle low-gravity motion. Fine lunar dust lifts and drifts outward with each movement, floating in slow arcs before settling back onto the ground. The astronaut pushes free in a deliberate, weightless motion, small fragments of the egg tumbling and spinning through the air. In the background, the deep darkness of space subtly shifts as stars glide with the camera's movement, emphasizing vast depth and scale. The camera performs a smooth, cinematic slow push-in, with natural parallax between the foreground dust, the astronaut, and the distant starfield. Ultra-realistic detail, physically accurate low-gravity motion, cinematic lighting, and a breath-taking, movie-like shot."
negative_prompt = "shaky, glitchy, low quality, worst quality, deformed, distorted, disfigured, motion smear, motion artifacts, fused fingers, bad anatomy, weird hand, ugly, transition, static."

frame_rate = 24.0
video, audio = pipe(
    image=image,
    prompt=prompt,
    negative_prompt=negative_prompt,
    width=768,
    height=512,
    num_frames=121,
    frame_rate=frame_rate,
    num_inference_steps=40,
    guidance_scale=4.0,
    output_type="np",
    return_dict=False,
)
video = (video * 255).round().astype("uint8")
video = torch.from_numpy(video)

encode_video(
    video[0],
    fps=frame_rate,
    audio=audio[0].float().cpu(),
    audio_sample_rate=pipe.vocoder.config.output_sampling_rate,
    output_path="ltx2_sample.mp4",
)
```

Note that LTX 2.0 video generation uses a lot of memory; it is necessary to use CPU offloading even for an A100 which has 80 GB VRAM (assuming no other memory optimizations other than `bf16` inference are used).

Here is an I2V sample from the above:

https://github.com/user-attachments/assets/1fedf731-0905-41b2-be0b-5531f4e4b8de

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu
@sayakpaul
@ofirbb

### CODE REVIEW

File: scripts/ltx2_test_full_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import argparse
```
Comment (yiyixuxu) [inline scripts/ltx2_test_full_pipeline.py]: let's remove these files

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 2]
```diff
@@ -0,0 +1,1360 @@
...
+    out = out.reshape(*out.shape[:-2], last)
+
+    if needs_reshape:
+        out = out.swapaxes(1, 2).reshape(b, t, -1)
+
+    out = out.to(dtype=x_dtype)
+    return out
+
+
+ROTARY_FN_MAP = {"interleaved": apply_interleaved_rotary_emb, "split": apply_split_rotary_emb}
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_ltx2.py]:
```suggestion
```
we can just do a `if ... else`

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 3]
```diff
@@ -0,0 +1,1360 @@
...
+
+        query = attn.to_q(hidden_states)
+        key = attn.to_k(encoder_hidden_states)
+        value = attn.to_v(encoder_hidden_states)
+
+        query = attn.norm_q(query)
+        key = attn.norm_k(key)
+
+        if query_rotary_emb is not None:
+            query = ROTARY_FN_MAP[attn.rope_type](query, query_rotary_emb)
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_ltx2.py]:
```suggestion
            if attn.rope_type === "interleaved"...
                query = apply_interleaved_rotary_emb(query, query_rotary_emb)
                key = apply_interleaved_rotary_emb(...)
            elif attn.rope_type == "split":
                ...
```

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 4]
```diff
@@ -0,0 +1,1360 @@
...
+        ca_video_rotary_emb: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        ca_audio_rotary_emb: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        audio_encoder_attention_mask: Optional[torch.Tensor] = None,
+        a2v_cross_attention_mask: Optional[torch.Tensor] = None,
+        v2a_cross_attention_mask: Optional[torch.Tensor] = None,
+        use_video_self_attn: bool = True,
+        use_audio_self_attn: bool = True,
+        use_a2v_cross_attn: bool = True,
+        use_v2a_cross_attn: bool = True,
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_ltx2.py]:
```suggestion
```
I don't see this arg is used , they seem always to be True 

can we remove them, along with all the `if` path inside `forward` related to them, to simplify the code here a bit

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 5]
```diff
@@ -0,0 +1,1360 @@
...
+        coords: Optional[torch.Tensor] = None,
+        batch_size: Optional[int] = None,
+        num_frames: Optional[int] = None,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        fps: float = 25.0,
+        shift: int = 0,
+        device: Optional[Union[str, torch.device]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if coords is not None:
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_ltx2.py]: i think coords is always pre-computed and passed here, no? 
can we simplify the code a bit, and remove all the code here to compute coords

File: src/diffusers/pipelines/ltx2/pipeline_ltx2.py [hunk 6]
```diff
@@ -0,0 +1,1125 @@
...
+else:
+    XLA_AVAILABLE = False
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import LTXPipeline
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2.py]: don't forget to update docstring example

File: src/diffusers/pipelines/ltx2/pipeline_ltx2.py [hunk 7]
```diff
@@ -0,0 +1,1125 @@
...
+    def attention_kwargs(self):
+        return self._attention_kwargs
+
+    @property
+    def interrupt(self):
+        return self._interrupt
+
+    @torch.no_grad()
+    @replace_example_docstring(EXAMPLE_DOC_STRING)
+    def __call__(
```
Comment (JoeGaffney) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2.py:744]: Hey has v2 lost all the condition support inputs LTXVideoCondition

I thought that was quite a nice system in v1, although v1 quality was not the best

Comment (sayakpaul) [reply src/diffusers/pipelines/ltx2/pipeline_ltx2.py:744]: Yes, we will ship that. But if you want to help us in shipping it in a separate PR, that is welcome as well.

File: src/diffusers/models/transformers/transformer_ltx2.py (hunk 4 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/transformer_ltx2.py]: Done in c39f1b87a

File: src/diffusers/pipelines/ltx2/pipeline_ltx2.py (hunk 6 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/ltx2/pipeline_ltx2.py]: Done bdcf23ec1

**Comment by sayakpaul** [comment]
Cc: @matanby if you want to test this PR on your end. We will shortly be adding the upsampling pipeline as well.

File: src/diffusers/models/transformers/transformer_ltx2.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/transformer_ltx2.py]: @dg845 WDYT?

**Comment by bghira** [comment]
no audio encode?

File: src/diffusers/models/transformers/transformer_ltx2.py (hunk 5 above)
Comment (dg845) [reply src/diffusers/models/transformers/transformer_ltx2.py]: I think this change is fine.

**Comment by dg845** [comment]
@bghira, so I understand correctly, is the request for an analogue of `diffusers.pipeline.ltx2.export_utils.encode_video` that only encodes the audio? `encode_video` should be able to create videos with audio.

**Comment by bghira** [comment]
the audio autoencoder is missing encode() function which exists in the LTX-2 repo from Lightricks, and ComfyUI is having audio encoding as well

**Comment by dg845** [comment]
@bghira thanks for the clarification! We will support the audio VAE encoder.

File: tests/models/transformers/test_models_transformer_ltx2.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+            "encoder_hidden_states": encoder_hidden_states,
+            "audio_encoder_hidden_states": audio_encoder_hidden_states,
+            "timestep": timestep,
+            "encoder_attention_mask": encoder_attention_mask,
+            "num_frames": num_frames,
+            "height": height,
+            "width": width,
+            "audio_num_frames": audio_num_frames,
+            "fps": 25.0,
+        }
```
Comment (sayakpaul) [inline tests/models/transformers/test_models_transformer_ltx2.py:75]: @dg845 should we not also pass the coordinates for RoPE?

Comment (dg845) [reply tests/models/transformers/test_models_transformer_ltx2.py:75]: The RoPE coordinates will still be calculated inside `LTX2VideoTransformer3DModel.forward` if they are not directly supplied to `forward`:

https://github.com/huggingface/diffusers/blob/5269ee5067f29ff89029963532821217890710e1/src/diffusers/models/transformers/transformer_ltx2.py#L1175-L1182

File: tests/models/autoencoders/test_models_autoencoder_kl_ltx2_audio.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Overriding as output shape is not the same as input shape for LTX 2.0 audio VAE
+    def test_output(self):
+        super().test_output(expected_output_shape=(2, 2, 5, 16))
+
+    @unittest.skip("Unsupported test.")
+    def test_outputs_equivalence(self):
+        pass
+
+    @unittest.skip("AutoencoderKLLTXAudio does not support `norm_num_groups` because it does not use GroupNorm.")
```
Comment (sayakpaul) [inline tests/models/autoencoders/test_models_autoencoder_kl_ltx2_audio.py]:
```suggestion
    @unittest.skip("AutoencoderKLLTX2Audio does not support `norm_num_groups` because it does not use GroupNorm.")
```

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 10]
```diff
@@ -0,0 +1,1321 @@
...
+            max_positions = (self.base_num_frames, self.base_height, self.base_width)
+        elif self.modality == "audio":
+            max_positions = (self.base_num_frames,)
+        # [B, num_pos_dims, num_patches] --> [B, num_patches, num_pos_dims]
+        grid = torch.stack([coords[:, i] / max_positions[i] for i in range(num_pos_dims)], dim=-1).to(device)
+        # Number of spatiotemporal dimensions (3 for video, 1 for audio and cross attn) times 2 for cos, sin
+        num_rope_elems = num_pos_dims * 2
+
+        # 3. Create a 1D grid of frequencies for RoPE
+        freqs_dtype = torch.float64 if self.double_precision else torch.float32
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_ltx2.py:819]: (nit): we could keep the `self.freqs_dtype` inside the init to skip doing it multiple times.

File: src/diffusers/models/transformers/transformer_ltx2.py [hunk 11]
```diff
@@ -0,0 +1,1321 @@
...
+                batch_size, audio_num_frames, audio_hidden_states.device, fps=fps
+            )
+
+        video_rotary_emb = self.rope(video_coords, device=hidden_states.device)
+        audio_rotary_emb = self.audio_rope(audio_coords, device=audio_hidden_states.device)
+
+        video_cross_attn_rotary_emb = self.cross_attn_rope(video_coords[:, 0:1, :], device=hidden_states.device)
+        audio_cross_attn_rotary_emb = self.cross_attn_audio_rope(
+            audio_coords[:, 0:1, :], device=audio_hidden_states.device
+        )
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_ltx2.py:1219]: (nit): would be nice to have a comment about the small indexing going on there.

File: src/diffusers/models/autoencoders/autoencoder_kl_ltx2_audio.py [hunk 12]
```diff
@@ -0,0 +1,805 @@
...
+            decoded_output = F.pad(decoded_output, padding)
+
+        decoded_output = decoded_output[:, :target_channels, :target_time, :target_freq]
+
+        return decoded_output
+
+
+class AutoencoderKLLTX2Audio(ModelMixin, AutoencoderMixin, ConfigMixin):
+    r"""
+    LTX2 audio VAE. Currently, only implements the decoder.
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/autoencoder_kl_ltx2_audio.py]:
```suggestion
    LTX2 audio VAE for encoding and decoding audio latent representations.
```

File: src/diffusers/models/autoencoders/autoencoder_kl_ltx2_audio.py [hunk 13]
```diff
@@ -0,0 +1,805 @@
...
+        sample_posterior: bool = False,
+        return_dict: bool = True,
+        generator: Optional[torch.Generator] = None,
+    ) -> Union[DecoderOutput, torch.Tensor]:
+        posterior = self.encode(sample).latent_dist
+        if sample_posterior:
+            z = posterior.sample(generator=generator)
+        else:
+            z = posterior.mode()
+        print(f"z shape: {z.shape}")
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/autoencoder_kl_ltx2_audio.py]: Remove.

File: docs/source/en/api/models/autoencoderkl_audio_ltx_2.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+from diffusers import AutoencoderKLLTX2Audio
+
+vae = AutoencoderKLLTX2Audio.from_pretrained("Lightricks/LTX-2", subfolder="vae", torch_dtype=torch.float32).to("cuda")
+```
+
+## AutoencoderKLLTX2Audio
+
+[[autodoc]] AutoencoderKLLTX2Audio
+    - decode
+    - all
```
Comment (sayakpaul) [inline docs/source/en/api/models/autoencoderkl_audio_ltx_2.md:29]:
```suggestion
    - encode
    - decode
    - all
```

**Review by sayakpaul — COMMENTED** [review]
Some small comments.

**Comment by dg845** [comment]
Merging as the CI failures are unrelated.

**Comment by hannalaguilar** [comment]
> # What does this PR do?
> This PR adds pipelines for the LTX 2.0 video generation model ([code](https://github.com/Lightricks/LTX-2), [weights](https://huggingface.co/Lightricks/LTX-2)). LTX 2.0 is an audio-video foundation model that generates videos with synced audio; it supports generation tasks such as text-to-video (T2V), text-image-to-video (TI2V), and more.
> 
> An example usage script for I2V is as follows:
> 
> ```python
> import torch
> from diffusers.pipelines.ltx2 import LTX2ImageToVideoPipeline
> from diffusers.pipelines.ltx2.export_utils import encode_video
> from diffusers.utils import load_image
> 
> pipe = LTX2ImageToVideoPipeline.from_pretrained("Lightricks/LTX-2", torch_dtype=torch.bfloat16)
> pipe.enable_model_cpu_offload()
> 
> prompt = "An astronaut hatches from a fragile egg on the surface of the Moon, the shell cracking and peeling apart in gentle low-gravity motion. Fine lunar dust lifts and drifts outward with each movement, floating in slow arcs before settling back onto the ground. The astronaut pushes free in a deliberate, weightless motion, small fragments of the egg tumbling and spinning through the air. In the background, the deep darkness of space subtly shifts as stars glide with the camera's movement, emphasizing vast depth and scale. The camera performs a smooth, cinematic slow push-in, with natural parallax between the foreground dust, the astronaut, and the distant starfield. Ultra-realistic detail, physically accurate low-gravity motion, cinematic lighting, and a breath-taking, movie-like shot."
> negative_prompt = "shaky, glitchy, low quality, worst quality, deformed, distorted, disfigured, motion smear, motion artifacts, fused fingers, bad anatomy, weird hand, ugly, transition, static."
> 
> frame_rate = 24.0
> video, audio = pipe(
> 	prompt=prompt,
> 	negative_prompt=negative_prompt,
> 	width=768,
> 	height=512,
> 	num_frames=121,
> 	frame_rate=frame_rate,
> 	num_inference_steps=40,
> 	guidance_scale=4.0,
> 	output_type="np",
> 	return_dict=False,
> )
> video = (video * 255).round().astype("uint8")
> video = torch.from_numpy(video)
> 
> encode_video(
> 	video[0],
> 	fps=frame_rate,
> 	audio=audio[0].float().cpu(),
> 	audio_sample_rate=pipe.vocoder.config.output_sampling_rate,
> 	output_path="ltx2_sample.mp4",
> )
> ```
> 
> Note that LTX 2.0 video generation uses a lot of memory; it is necessary to use CPU offloading even for an A100 which has 80 GB VRAM (assuming no other memory optimizations other than `bf16` inference are used).
> 
> Here is an I2V sample from the above:
> 
>  ltx2_i2v_sample.mp4 
> ## Who can review?
> Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.
> 
> @yiyixuxu @sayakpaul @ofirbb

In this example, an `image is not being passed to the pipeline. Should be:

```
image = load_image("image.png")
video, audio = pipe(
	prompt=prompt,
	negative_prompt=negative_prompt,
        image=image,
	width=768,
	height=512,
	num_frames=121,
	frame_rate=frame_rate,
	num_inference_steps=40,
	guidance_scale=4.0,
	output_type="np",
	return_dict=False,
)
```
